### DISCUSSION TITLE
How to deploy MPT with llama.cpp?

### DISCUSSION THREAD

**Discussion by streetycat** [abetlen/llama-cpp-python#939] [Q&A]
I found that llama.cpp already supports mpt, I downloaded [gguf](https://huggingface.co/maddes8cht/mosaicml-mpt-30b-chat-gguf/resolve/main/mosaicml-mpt-30b-chat-Q8_0.gguf) from here, and it did load it with llama.cpp, but its return result looks bad.

I start the server as follow:
```
git clone https://github.com/abetlen/llama-cpp-python.git
cd llama-cpp-python
docker build -t llama-cpp-python-cuda docker/cuda_simple/
docker run --gpus all --rm -it -p 8000:8000 -v ${/path/to/models}:/models -e MODEL=/models/${model-filename} llama-cpp-python-cuda python3 -m llama_cpp.server --n_gpu_layers ${X} --n_ctx ${L}
```

And I post the request as follow:

```
URL: http://localhost:8000/v1/chat/completions
BODY:

{
"messages": [
{"role": "user", "content": "What is 5+7?"}],
"max_tokens": 8000
}

RESPONCE:

{
    "id": "chatcmpl-bfe9eaf4-2ab4-419d-ab8b-313add0706f9",
    "object": "chat.completion",
    "created": 1700798859,
    "model": "/models/mosaicml-mpt-30b-chat-Q8_0.gguf",
    "choices": [
        {
            "index": 0,
            "message": {
                "content": "></s>\n<s>&nbsp;</s><br/> </body> </html> \n\nNotice the use of JavaScript to handle user input and submit it via AJAX. ...\n\n",
                "role": "assistant"
            },
            "finish_reason": "stop"
        }
    ],
    "usage": {
        "prompt_tokens": 16,
        "completion_tokens": 2258,
        "total_tokens": 2274
    }
}
```

But, it works well on the demo page:
![1700799966235](https://github.com/ggerganov/llama.cpp/assets/34021271/3c1bf567-d3c3-4625-9212-c4817ced5b92)

I worked well with llama in the same method, I don't know what's wrong happened.

**Comment by gardner** [comment]
How are you sending the POST? Can you share your `curl` command or `fetch()` request?

**Comment by streetycat** [comment]
Thank you, I have found a parameter(`chat_format`) in source code for startup to fix it.

![image](https://github.com/abetlen/llama-cpp-python/assets/34021271/edbde252-0ce2-4e3e-9339-993fad5a2b05)

and update the command:
```
docker run --gpus all --rm -it -p 8000:8000 -v ${/path/to/models}:/models -e MODEL=/models/${model-filename} llama-cpp-python-cuda python3 -m llama_cpp.server --n_gpu_layers ${X} --n_ctx ${L} --chat_format mistrallite
```

the responce:

![image](https://github.com/abetlen/llama-cpp-python/assets/34021271/ef8ce089-c928-422a-b551-098abbc68cd6)


By the way, maybe the `</s>` is a bug for the `chat_format`?

**Reply by gardner** [reply]
I can't find anywhere that mosaic describes the prompt template. There is one note on the model card here: https://huggingface.co/TheBloke/mpt-30B-chat-GGML#prompt-template

**Reply by streetycat** [reply]
Thank you for your help, I am a beginner and don't know where to find this information. I guess gguf and ggml are just an information organization format, and the specific behavior of the model should be consistent with the official mpt.

**Reply by gardner** [reply]
Correct. `.gguf` is just a file format to store the model weights. Much like `.zip` is a file format to store other files. The contents of a `.gguf` can contain different models that have been trained to use different prompt formats. You can usually find the prompt format on the model card on HuggingFace but sometimes you need to do some searching to figure out what to use. 

Good luck in your learnings!

**Comment by tk-master** [comment]
Definitely looks like prompt format issue
